### PR TITLE
Azure Recovery Services VM Backup Policy

### PR DESCRIPTION
**Pull request by suyeb786**
##### SUMMARY

This module introducing the VM Backup Policy features listed below
1. Create VM Backup Policy
2. Delete VM Backup Policy
3. Get VM Backup Policy Details

#255 
##### ISSUE TYPE

- New Module Pull Request

##### COMPONENT NAME

azure_rm_vmbackuppolicy
azure_rm_vmbackuppolicy_info
##### ADDITIONAL INFORMATION

```paste below

```

### CODE REVIEW

**Comment by suyeb786** [comment]
@Fred-sun .
Any feedback on this.

**Comment by Fred-sun** [comment]
@suyeb786 I haven't reviewed this PR recently because I am busy with other things. I will arrange time to review it as soon as possible. Thank you!

File: plugins/modules/azure_rm_vmbackuppolicy.py [hunk 1]
```diff
@@ -0,0 +1,463 @@
...
+        required: true
+        type: str
+    vault_name:
+        description:
+            - Recovery Service Vault Name.
+        required: true
+        type: str
+    time:
+        description:
+            - Retention times of retention policy in UTC.
```
Comment (Fred-sun) [inline plugins/modules/azure_rm_vmbackuppolicy.py]: trailing whitespace

File: plugins/modules/azure_rm_vmbackuppolicy.py [hunk 2]
```diff
@@ -0,0 +1,463 @@
...
+     weeks: ['First', 'Fourth']
+     months: ['February', 'November']
+     count: 4
+     state: present
+- name: Delete VM Backup Policy
+  azure_rm_backvmuppolicy:
+     name: 'myBackupPolicy'
+     vault_name: 'myVault'
+     resource_group: 'myResourceGroup'
+     state: absent
```
Comment (Fred-sun) [inline plugins/modules/azure_rm_vmbackuppolicy.py]: trailing whitespace

File: plugins/modules/azure_rm_vmbackuppolicy.py [hunk 3]
```diff
@@ -0,0 +1,463 @@
...
+                        ],
+                        "retentionDuration": {
+                          "count": 4,
+                          "durationType": "Years"
+                        }
+                      }
+                    },
+                    "timeZone": "Pacific Standard Time",
+                    "protectedItemsCount": 0
+                  }
```
Comment (Fred-sun) [inline plugins/modules/azure_rm_vmbackuppolicy.py]: trailing whitespace

File: plugins/modules/azure_rm_vmbackuppolicy_info.py [hunk 4]
```diff
@@ -0,0 +1,266 @@
...
+    - azure.azcollection.azure_tags
+author:
+    - Suyeb Ansari (@suyeb786)
+'''
+
+EXAMPLES = '''
+  azure_rm_backvmuppolicy_info:
+     name: 'myBackupPolicy'
+     vault_name: 'myVault'
+     resource_group: 'myResourceGroup'
```
Comment (Fred-sun) [inline plugins/modules/azure_rm_vmbackuppolicy_info.py]: trailing whitespace

File: plugins/modules/azure_rm_vmbackuppolicy_info.py [hunk 5]
```diff
@@ -0,0 +1,266 @@
...
+                        ],
+                        "retentionDuration": {
+                          "count": 4,
+                          "durationType": "Years"
+                        }
+                      }
+                    },
+                    "timeZone": "Pacific Standard Time",
+                    "protectedItemsCount": 0
+                  }
```
Comment (Fred-sun) [inline plugins/modules/azure_rm_vmbackuppolicy_info.py]: trailing whitespace

File: plugins/modules/azure_rm_vmbackuppolicy_info.py [hunk 6]
```diff
@@ -0,0 +1,266 @@
...
+        self.url = None
+        self.status_code = [200, 202]
+
+        self.query_parameters = {}
+        self.query_parameters['api-version'] = '2019-05-13'
+        self.header_parameters = {}
+        self.header_parameters['Content-Type'] = 'application/json; charset=utf-8'
+
+        super(BackupPolicyVMInfo, self).__init__(derived_arg_spec=self.module_arg_spec,
+                                             supports_check_mode=True,
```
Comment (Fred-sun) [inline plugins/modules/azure_rm_vmbackuppolicy_info.py]: continuation line under-indented for visual indent ( include line 205,2.6,207)

File: plugins/modules/azure_rm_vmbackuppolicy_info.py [hunk 7]
```diff
@@ -0,0 +1,266 @@
...
+        except CloudError as e:
+            self.log('Backup policy does not exist.')
+            self.fail('Error in fetching VM Backup Policy {0}'.format(str(e)))
+        try:
+            response = json.loads(response.text)
+        except Exception:
+            response = {'text': response.text}
+
+        return response
+
```
Comment (Fred-sun) [inline plugins/modules/azure_rm_vmbackuppolicy_info.py:260]: expected 2 blank lines

File: plugins/modules/azure_rm_vmbackuppolicy.py [hunk 8]
```diff
@@ -0,0 +1,463 @@
...
+               + '/vaults' + '/' \
+               + self.vault_name + '/' \
+               + "backupPolicies/" \
+               + self.name
+
+    def set_schedule_run_time(self):
+        return time.strftime("%Y-%m-%d", time.gmtime()) + "T" + self.time + ":00Z"
+
+    def get_body(self):
+        self.log('backup attributes {}'.format(self.body))
```
Comment (Fred-sun) [inline plugins/modules/azure_rm_vmbackuppolicy.py]:
```suggestion
        self.log('backup attributes {0}'.format(self.body))
```

**Review by Fred-sun — COMMENTED** [review]
The parameters count, months, time, weekdays, weeks of plugins/modules/azure_rm_vmbackuppolicy.py do not have default values ​​and require, but they are defined in argument_spec, and months, weekdays, weeks are not defined in the document.

File: tests/integration/targets/azure_rm_vmbackuppolicy/tasks/main.yml [hunk 9]
```diff
@@ -0,0 +1,56 @@
+- name: Fix resource prefix
+  set_fact:
+    name: "myPolicy{{ resource_group | hash('md5') | truncate(22, True, '') }}"
+    vault_name: "myVault"
+
```
Comment (Fred-sun) [inline tests/integration/targets/azure_rm_vmbackuppolicy/tasks/main.yml:14]: I think the recovery service vault should be created in this sample and deleted at the end.  Please add create recovery vault service at this place.

File: tests/integration/targets/azure_rm_vmbackuppolicy/tasks/main.yml [hunk 10]
```diff
@@ -0,0 +1,56 @@
...
+    resource_group: "{{ resource_group }}"
+    name: "{{ name }}"
+    vault_name: "{{ vault_name }}"
+    state: "absent"
+  register: output
+
+- name: Assert that output has changed
+  assert:
+    that:
+      - output.changed
```
Comment (Fred-sun) [inline tests/integration/targets/azure_rm_vmbackuppolicy/tasks/main.yml:78]: Please add delete recovery vault service at this place.

File: tests/integration/targets/azure_rm_vmbackuppolicy/tasks/main.yml (hunk 9 above)
Comment (suyeb786) [reply tests/integration/targets/azure_rm_vmbackuppolicy/tasks/main.yml:14]: test cases are added

**Comment by suyeb786** [comment]
Hello @Fred-sun ,
Any update on this.

**Comment by Fred-sun** [comment]
@suyeb786 Please help resolve the conflicting files, I will complete the test as soon as possible and recommend merging! Thank you!

**Comment by suyeb786** [comment]
Hi @Fred-sun ,
The conflict has been resolved.

**Comment by Fred-sun** [comment]
@suyeb786  The parameters count, months, time, weekdays, weeks of plugins/modules/azure_rm_vmbackuppolicy.py do not have default values ​​and require, but they are defined in argument_spec, and months, weekdays, weeks are not defined type in the document.

File: tests/integration/targets/azure_rm_vmbackuppolicy/tasks/main.yml [hunk 11]
```diff
@@ -0,0 +1,84 @@
...
+    name: "{{ name }}"
+    location: "{{ location }}"
+    state: "present"
+  register: vault
+
+- name: Assert ssucceeded and results include vault name
+  assert:
+    that:
+      - vault.changed
+      - vault.state.name is defined
```
Comment (Fred-sun) [inline tests/integration/targets/azure_rm_vmbackuppolicy/tasks/main.yml]:
```suggestion

```

File: plugins/modules/azure_rm_vmbackuppolicy.py [hunk 12]
```diff
@@ -0,0 +1,468 @@
...
+        required: true
+        type: str
+    vault_name:
+        description:
+            - Recovery Service Vault Name.
+        required: true
+        type: str
+    time:
+        description:
+            - Retention times of retention policy in UTC.
```
Comment (Fred-sun) [inline plugins/modules/azure_rm_vmbackuppolicy.py]:
```suggestion
            - Retention times of retention policy in UTC.
```

File: plugins/modules/azure_rm_vmbackuppolicy.py [hunk 13]
```diff
@@ -0,0 +1,468 @@
...
+     weeks: ['First', 'Fourth']
+     months: ['February', 'November']
+     count: 4
+     state: present
+- name: Delete VM Backup Policy
+  azure_rm_backvmuppolicy:
+     name: 'myBackupPolicy'
+     vault_name: 'myVault'
+     resource_group: 'myResourceGroup'
+     state: absent
```
Comment (Fred-sun) [inline plugins/modules/azure_rm_vmbackuppolicy.py]:
```suggestion
     state: absent
```

File: plugins/modules/azure_rm_vmbackuppolicy.py [hunk 14]
```diff
@@ -0,0 +1,468 @@
...
+                        ],
+                        "retentionDuration": {
+                          "count": 4,
+                          "durationType": "Years"
+                        }
+                      }
+                    },
+                    "timeZone": "Pacific Standard Time",
+                    "protectedItemsCount": 0
+                   }
```
Comment (Fred-sun) [inline plugins/modules/azure_rm_vmbackuppolicy.py]:
```suggestion
                   }
```

File: plugins/modules/azure_rm_vmbackuppolicy_info.py [hunk 15]
```diff
@@ -0,0 +1,267 @@
...
+    - azure.azcollection.azure_tags
+author:
+    - Suyeb Ansari (@suyeb786)
+'''
+
+EXAMPLES = '''
+   azure_rm_backvmuppolicy_info:
+     name: 'myBackupPolicy'
+     vault_name: 'myVault'
+     resource_group: 'myResourceGroup'
```
Comment (Fred-sun) [inline plugins/modules/azure_rm_vmbackuppolicy_info.py]:
```suggestion
     resource_group: 'myResourceGroup'
```

File: plugins/modules/azure_rm_vmbackuppolicy_info.py [hunk 16]
```diff
@@ -0,0 +1,267 @@
...
+                        ],
+                        "retentionDuration": {
+                          "count": 4,
+                          "durationType": "Years"
+                        }
+                      }
+                    },
+                    "timeZone": "Pacific Standard Time",
+                    "protectedItemsCount": 0
+                   }
```
Comment (Fred-sun) [inline plugins/modules/azure_rm_vmbackuppolicy_info.py]:
```suggestion
                   }
```

File: plugins/modules/azure_rm_vmbackuppolicy.py [hunk 17]
```diff
@@ -0,0 +1,468 @@
...
+    vault_name:
+        description:
+            - Recovery Service Vault Name.
+        required: true
+        type: str
+    time:
+        description:
+            - Retention times of retention policy in UTC. 
+        required: false
+        default: 12:00
```
Comment (Fred-sun) [inline plugins/modules/azure_rm_vmbackuppolicy.py]:
```suggestion
        default: '12:00'
```

**Comment by suyeb786** [comment]
@Fred-sun any update?

**Comment by Fred-sun** [comment]
> @Fred-sun any update?

Please add a line in the file below---"plugins/modules/azure_rm_vmbackuppolicy.py validate-modules:parameter-list-no-elements"
```e.g.
tests/sanity/ignore-2.10.txt
tests/sanity/ignore-2.11.txt

**Comment by suyeb786** [comment]
@Fred-sun 
tests/sanity/ignore-2.10.txt and tests/sanity/ignore-2.11.txt files have been updated with the following line.
plugins/modules/azure_rm_vmbackuppolicy.py validate-modules:parameter-list-no-elements

**Comment by Fred-sun** [comment]
> @Fred-sun
> tests/sanity/ignore-2.10.txt and tests/sanity/ignore-2.11.txt files have been updated with the following line.
> plugins/modules/azure_rm_vmbackuppolicy.py validate-modules:parameter-list-no-elements

Because there is something wrong with azure_rm_recoveryservicesvault, please delete lines 15~19 temporarily， This should be fine! Thank you very much!

**Comment by suyeb786** [comment]
@Fred-sun 
line number 15~19 temporarily from which file should I delete?
Please help me with this.

**Comment by Fred-sun** [comment]
> @Fred-sun
> line number 15~19 temporarily from which file should I delete?
> Please help me with this.

tests/integration/targets/azure_rm_vmbackuppolicy/tasks/main.yml , Thanks!

**Comment by suyeb786** [comment]
Hello @Fred-sun 
tasks/main.yml file is updated.
Thank you

File: tests/integration/targets/azure_rm_vmbackuppolicy/tasks/main.yml [hunk 18]
```diff
@@ -0,0 +1,78 @@
...
+    name: "{{ name }}"
+    location: "{{ location }}"
+    state: "present"
+  register: vault
+
+- name: Create VM Backup Policy
+  azure_rm_vmbackuppolicy:
+    resource_group: "{{ resource_group }}"
+    name: "{{ name }}"
+    vault_name: "{{ vault.state.name }}"
```
Comment (Fred-sun) [inline tests/integration/targets/azure_rm_vmbackuppolicy/tasks/main.yml]:
```suggestion
    vault_name: "{{ vault.response.name }}"
```

File: tests/integration/targets/azure_rm_vmbackuppolicy/tasks/main.yml [hunk 19]
```diff
@@ -0,0 +1,78 @@
...
+- name: Assert that output has changed
+  assert:
+    that:
+      - output.changed
+      
+- name: Create VM Backup Policy (idempotent)
+  azure_rm_vmbackuppolicy:
+    resource_group: "{{ resource_group }}"
+    name: "{{ name }}"
+    vault_name: "{{ vault.state.name }}"
```
Comment (Fred-sun) [inline tests/integration/targets/azure_rm_vmbackuppolicy/tasks/main.yml]:
```suggestion
    vault_name: "{{ vault.response.name }}"
```

File: tests/integration/targets/azure_rm_vmbackuppolicy/tasks/main.yml [hunk 20]
```diff
@@ -0,0 +1,78 @@
...
+- name: Assert that output has no changed
+  assert:
+    that:
+      - not output.changed
+
+- name: Get VM Backup Policy Details
+  azure_rm_vmbackuppolicy_info:
+    resource_group: "{{ resource_group }}"
+    name: "{{ name }}"
+    vault_name: "{{ vault.state.name }}"
```
Comment (Fred-sun) [inline tests/integration/targets/azure_rm_vmbackuppolicy/tasks/main.yml]:
```suggestion
    vault_name: "{{ vault.response.name }}"
```

File: tests/integration/targets/azure_rm_vmbackuppolicy/tasks/main.yml [hunk 21]
```diff
@@ -0,0 +1,78 @@
...
+  assert:
+    that:
+      - output.response.id != None
+      - output.response.name != None
+
+- name: Delete VM Backup Policy
+  azure_rm_vmbackuppolicy:
+    resource_group: "{{ resource_group }}"
+    name: "{{ name }}"
+    vault_name: "{{ vault.state.name }}"
```
Comment (Fred-sun) [inline tests/integration/targets/azure_rm_vmbackuppolicy/tasks/main.yml]:
```suggestion
    vault_name: "{{ vault.response.name }}"
```

File: tests/integration/targets/azure_rm_vmbackuppolicy/tasks/main.yml [hunk 22]
```diff
@@ -0,0 +1,78 @@
...
+
+- name: Assert that output has changed
+  assert:
+    that:
+      - output.changed
+
+- name: Delete Azure Recovery Service vault
+  azure_rm_recoveryservicesvault:
+    resource_group: "{{ resource_group }}"
+    name: "{{ vault.state.name }}"
```
Comment (Fred-sun) [inline tests/integration/targets/azure_rm_vmbackuppolicy/tasks/main.yml]:
```suggestion
    name: "{{ vault.response.name }}"
```

**Pull request merged**